File: days/day_11/solutions/solution.hs
import System.IO
import Data.Char
import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map
import Data.List

-- Import utilities
getInputPath :: Int -> FilePath
getInputPath _ = "../data/input.txt"

readInput :: FilePath -> IO String
readInput filePath = readFile filePath

trim :: String -> String
trim = reverse . dropWhile isSpace . reverse . dropWhile isSpace

splitOn :: Char -> String -> [String]
splitOn _ [] = []
splitOn delimiter str =
  case break (== delimiter) str of
    (before, []) -> [before]
    (before, _:after) -> before : splitOn delimiter after

buildGraph :: [String] -> Map String [String]
buildGraph inputLines = foldl' buildGraph' Map.empty inputLines
  where
    buildGraph' graph line =
      let parts = splitOn ':' line
      in if length parts == 2
         then let device = trim $ head parts
                  outputs_str = trim $ parts !! 1
                  outputs = if null outputs_str
                           then []
                           else words outputs_str
              in Map.insert device outputs graph
         else graph

countPathsPart1 :: String -> Map String [String] -> Map String Integer -> (Integer, Map String Integer)
countPathsPart1 node graph memo =
  if node == "out"
  then (1, memo)
  else case Map.lookup node memo of
    Just value -> (value, memo)
    Nothing ->
      let neighbors = Map.findWithDefault [] node graph
          (count, newMemo) = foldl' (\(acc, mem) neighbor ->
                                       let (val, updatedMem) = countPathsPart1 neighbor graph mem
                                       in (acc + val, updatedMem))
                                    (0, memo)
                                    neighbors
      in (count, Map.insert node count newMemo)

countPathsPart2 :: String -> Bool -> Bool -> Map String [String] -> Map (String, Bool, Bool) Integer -> (Integer, Map (String, Bool, Bool) Integer)
countPathsPart2 node visitedFFT visitedDAC graph memo =
  if node == "out"
  then (if visitedFFT && visitedDAC then 1 else 0, memo)
  else let key = (node, visitedFFT, visitedDAC)
       in case Map.lookup key memo of
         Just value -> (value, memo)
         Nothing ->
           let newVisitedFFT = visitedFFT || (node == "fft")
               newVisitedDAC = visitedDAC || (node == "dac")
               neighbors = Map.findWithDefault [] node graph
               (count, newMemo) = foldl' (\(acc, mem) neighbor ->
                                            let (val, updatedMem) = countPathsPart2 neighbor newVisitedFFT newVisitedDAC graph mem
                                            in (acc + val, updatedMem))
                                         (0, memo)
                                         neighbors
           in (count, Map.insert key count newMemo)

solve :: String -> (String, String)
solve inputData =
  let inputLines = filter (not . null . trim) $ Data.List.lines inputData
      graph = buildGraph inputLines
      
      part1Count = if Map.member "you" graph
                   then fst $ countPathsPart1 "you" graph Map.empty
                   else 0
      
      part2Count = if Map.member "svr" graph
                   then fst $ countPathsPart2 "svr" False False graph Map.empty
                   else 0
  in (show part1Count, show part2Count)

main :: IO ()
main = do
  inputData <- readInput "../data/input.txt"
  let (part1, part2) = solve inputData
  putStrLn $ "Part 1: " ++ part1
  putStrLn $ "Part 2: " ++ part2

File: days/day_11/solutions/solution.ex
defmodule Solution do
  def solve(input_data) do
    lines = String.trim(input_data) |> String.split("\n")

    # Build graph: device -> list of outputs
    graph = Enum.reduce(lines, %{}, fn line, acc ->
      case String.split(line, ":", parts: 2) do
        [device, outputs_str] ->
          device = String.trim(device)
          outputs_str = String.trim(outputs_str)
          outputs = if outputs_str == "", do: [], else: String.split(outputs_str, ~r/\s+/)
          Map.put(acc, device, outputs)
        _ ->
          acc
      end
    end)

    # Part 1: Count paths from "you" to "out"
    {part1_count, _} = if Map.has_key?(graph, "you") do
      count_paths_part1("you", graph, %{})
    else
      {0, %{}}
    end

    # Part 2: Count paths from "svr" to "out" that include both "fft" and "dac"
    {part2_count, _} = if Map.has_key?(graph, "svr") do
      count_paths_part2("svr", false, false, graph, %{})
    else
      {0, %{}}
    end

    {to_string(part1_count), to_string(part2_count)}
  end

  defp count_paths_part1(node, graph, memo) do
    if node == "out" do
      {1, memo}
    else
      case Map.fetch(memo, node) do
        {:ok, value} -> {value, memo}
        :error ->
          neighbors = Map.get(graph, node, [])
          {count, new_memo} = Enum.reduce(neighbors, {0, memo}, fn neighbor, {acc, mem} ->
            {val, updated_mem} = count_paths_part1(neighbor, graph, mem)
            {acc + val, updated_mem}
          end)
          final_memo = Map.put(new_memo, node, count)
          {count, final_memo}
      end
    end
  end

  defp count_paths_part2(node, visited_fft, visited_dac, graph, memo) do
    if node == "out" do
      result = if visited_fft && visited_dac, do: 1, else: 0
      {result, memo}
    else
      key = {node, visited_fft, visited_dac}
      case Map.fetch(memo, key) do
        {:ok, value} -> {value, memo}
        :error ->
          new_visited_fft = visited_fft || (node == "fft")
          new_visited_dac = visited_dac || (node == "dac")
          neighbors = Map.get(graph, node, [])
          {count, new_memo} = Enum.reduce(neighbors, {0, memo}, fn neighbor, {acc, mem} ->
            {val, updated_mem} = count_paths_part2(neighbor, new_visited_fft, new_visited_dac, graph, mem)
            {acc + val, updated_mem}
          end)
          final_memo = Map.put(new_memo, key, count)
          {count, final_memo}
      end
    end
  end
end

input_data = File.read!("../data/input.txt")
{part1, part2} = Solution.solve(input_data)
IO.puts("Part 1: #{part1}")
IO.puts("Part 2: #{part2}")
